[SimplifyCFG] Convert switch to cmp/select sequence

Acim-Maravic opened the pull request:
I have added the option -switch-to-select, which, when passed, will trigger this optimization. I am not sure where this code should reside and if I am on the right track. What I am currently attempting is to add support for simple switches, i.e., switches where all values in the PHI node are constants. Now, I am trying to add support for more complex switch cases, i.e., where the values in the PHI node do not have to be constants.

When I am getting something like this:

switch i32 %cond.freeze1, label %21 [
  i32 0, label %15
  i32 1, label %16
  i32 2, label %17
  i32 3, label %19
  ]

15:                                               ; preds = %6
  br label %21

16:                                               ; preds = %6
  br label %21

17:                                               ; preds = %6
  %18 = fneg reassoc nnan nsz arcp contract afn float %.i0
  br label %21

19:                                               ; preds = %6
  %20 = fneg reassoc nnan nsz arcp contract afn float %.i0
  br label %21

21:                                               ; preds = %6, %15, %16, %17, %19, %.entry
  %samplePos.1.i0 = phi float [ 0.000000e+00, %.entry ], [ 0.000000e+00, %6 ], [ %20, %19 ], [ 1.000000e+00, %17 ], [ %.i0, %16 ], [ 1.000000e+00, %15 ]
  %samplePos.1.i1 = phi float [ 0.000000e+00, %.entry ], [ 0.000000e+00, %6 ], [ %.i1, %19 ], [ %.i1, %17 ], [ %.i1, %16 ], [ 1.000000e+00, %15 ]
  %samplePos.1.i2 = phi float [ 0.000000e+00, %.entry ], [ 0.000000e+00, %6 ], [ -1.000000e+00, %19 ], [ %18, %17 ], [ 1.000000e+00, %16 ], [ 1.000000e+00, %15 ]


Next step is to gather all blocks that can be safely merged be merged into 21 in order for me to create selects and to allow non constant values in helpers.

jayfoad commented:
The general idea of doing this in SimplifyCFG seems reasonable to me. However I am not too familiar with the existing code in SimplifyCFG.

arsenm commented on llvm/lib/Transforms/Scalar/SimplifyCFGPass.cpp:
```diff
@@ -61,6 +61,10 @@ static cl::opt<bool> UserSwitchRangeToICmp(
     cl::desc(
         "Convert switches into an integer range comparison (default = false)"));
 
+static cl::opt<bool> UserSwitchToSelect(
+    "switch-to-select", cl::Hidden, cl::init(false),
+    cl::desc("Convert switches into icmp + select (default = false)"));
+
```
do we really need new cl::opts for this? Can just be a pass parameter?

Acim-Maravic replied:
This is just for testing purposes, I guess in one point we will instantiate a pass with enable by default...

tsymalla commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -5916,7 +5916,7 @@ static bool initializeUniqueCases(SwitchInst *SI, PHINode *&PHI,
 
     // Only one value per case is permitted.
-    if (Results.size() > 1)
+    if (Results.size() > 3) // How many PHI instructions are hendled
```
Typo: handled

tsymalla commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -5916,7 +5916,7 @@ static bool initializeUniqueCases(SwitchInst *SI, PHINode *&PHI,
       return false;
 
     // Only one value per case is permitted.
```
Comment needs update?

tsymalla commented on llvm/test/Transforms/SimplifyCFG/AMDGPU/switch-to-select.ll:
```diff
@@ -0,0 +1,312 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
```
Shouldn't there be a non-AMDGPU test case since this is a general optimization?

Acim-Maravic replied:
Yes, I will do it that way.

tsymalla commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -5967,6 +5986,22 @@ static Value *foldSwitchToSelect(const SwitchCaseResultVectorTy &ResultVector,
+    if (IsSizeOkay && ResultVector.size() > 2) {
+      Value *FinalSelect =
+          createSelectChain(Condition, DefaultResult, ResultVector, 0, Builder);
+      if (FinalSelect)
```
`createSelectChain` always returns a non-`nullptr` value?

Acim-Maravic replied:
Yes.

tsymalla commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -7028,7 +7071,11 @@ bool SimplifyCFGOpt::simplifySwitch(SwitchInst *SI, IRBuilder<> &Builder) {
-  if (trySwitchToSelect(SI, Builder, DTU, DL, TTI))
+  bool IsSwitchToSelect = false;
+  if (Options.ConvertSwitchToSelect)
+    IsSwitchToSelect = true;
```
`IsSwitchToSelect = Options.ConvertSwitchToSelect`?

Acim-Maravic replied:
Yes.

tsymalla commented on llvm/test/Transforms/SimplifyCFG/AMDGPU/switch-to-select.ll:
```diff
@@ -0,0 +1,312 @@
+; ALL-NEXT:    [[SWITCH_SELECTCMP1:%.*]] = icmp eq i32 [[COND_FREEZE1]], 1
+; ALL-NEXT:    [[SWITCH_SELECTCMP2:%.*]] = icmp eq i32 [[COND_FREEZE1]], 2
+; ALL-NEXT:    [[SWITCH_SELECTCMP3:%.*]] = icmp eq i32 [[COND_FREEZE1]], 3
+; ALL-NEXT:    [[SWITCHACIM_SELECT:%.*]] = select i1 [[SWITCH_SELECTCMP3]], float 4.000000e+00, float 0.000000e+00
```
Is the name `Switchacim` intended?

Acim-Maravic replied:
I have forgotten to update tests, it was just for debugging.

tsymalla commented on llvm/test/Transforms/SimplifyCFG/AMDGPU/switch-to-select.ll:
```diff
@@ -0,0 +1,312 @@
+; ALL-NEXT:    [[TMP20:%.*]] = fneg reassoc nnan nsz arcp contract afn float [[DOTI0]]
+; ALL-NEXT:    br label [[TMP21]]
+; ALL:       21:
+; ALL-NEXT:    [[SAMPLEPOS_1_I0:%.*]] = phi float [ 0.000000e+00, [[DOTENTRY:%.*]] ], [ 0.000000e+00, [[TMP6]] ], [ [[TMP20]], [[TMP19]] ], [ 1.000000e+00, [[TMP17]] ], [ [[DOTI0]], [[TMP16]] ], [ 1.000000e+00, [[TMP15]] ]
```
So, this is not working on partially constant phi nodes, right? It would be a great addition since we'd be able to generate selects out of code like shown in this test, and I guess this is something that will often appear in practice.

Acim-Maravic replied:
Now, it should be working with non-constant PHI Nodes, and with multiple PHI nodes.

arsenm commented on llvm/test/Transforms/SimplifyCFG/AMDGPU/switch-to-select.ll:
```diff
@@ -0,0 +1,312 @@
+  .entry:
+  %PerspInterpCenter.i1 = extractelement <2 x float> %PerspInterpCenter, i64 1
+  %PerspInterpCenter.i0 = extractelement <2 x float> %PerspInterpCenter, i64 0
+  %0 = call float @llvm.amdgcn.interp.p1(float %PerspInterpCenter.i0, i32 immarg 1, i32 immarg 0, i32 %PrimMask) #1
```
Use named values in tests

arsenm reviewed (commented):
Can you change the title? I have no idea what it means as-is

Acim-Maravic commented:
Fixed support for handling of multiple PHI nodes. Reworked implementation to try to carve out cases with constant values instead of giving up if not all cases are constant.

XChy commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6166,6 +6166,279 @@ static bool trySwitchToSelect(SwitchInst *SI, IRBuilder<> &Builder,
+  if (!IsDefault) {
+    for (Instruction &I : CaseDest->instructionsWithoutDebug(false)) {
+
+      if (I.isTerminator()) {
+        // If the terminator is a simple branch, continue to the next block.
+        if (I.getNumSuccessors() != 1 || I.isSpecialTerminator())
```
```suggestion
      if (BranchInst* BI = dyn_cast<BranchInst>(&I)) {
        // If the terminator is a simple branch, continue to the next block.
        if (!BI->isUnconditional())
```

XChy commented on llvm/test/Transforms/SimplifyCFG/switch-to-select.ll:
```diff
@@ -0,0 +1,341 @@
+return:                                               ; preds = %sw.bb1, %sw.bb2, %sw.bb3, %sw.bb4, %sw, %.entry
+  %samplePos.1.i0 = phi float [ 0.000000e+00, %.entry ], [ %20, %sw.bb4 ], [ -1.000000e+00, %sw.bb3 ], [ %.i0, %sw.bb2 ], [ 1.000000e+00, %sw.bb1 ], [ 4.000000e+00, %sw ]
+  ret float %samplePos.1.i0
+}
```
Negative tests? For example,
- No common dest for cases.
- The switch jumps direct into the dest block.
- Loop with switch.

tsymalla replied:
Can you please add these tests?

XChy reviewed (commented):
Generally, it's hard for me to understand the details of your code. Could you please add more comments to explain? And I think it's not always profitable to convert a big switch into a long select sequence. We need to add a threshold for it.

arsenm commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6166,6 +6166,279 @@ static bool trySwitchToSelect(SwitchInst *SI, IRBuilder<> &Builder,
+    SmallVector<std::pair<Value *, SmallVector<Value *, 4>>, 2>;
+
+using PHINodeToCaseEntryValueMapTy = std::map<
+    llvm::PHINode *,
```
Don't need llvm::

arsenm commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6166,6 +6166,279 @@ static bool trySwitchToSelect(SwitchInst *SI, IRBuilder<> &Builder,
+
+    return SelectedValue;
+
+  } else {
```
No else after return

arsenm commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6166,6 +6166,279 @@ static bool trySwitchToSelect(SwitchInst *SI, IRBuilder<> &Builder,
+      if (Index != -1) {
+        Value *Val = PHI->getIncomingValue(Index);
+        if (isa<Constant>(Val)) {
+          PHI->setIncomingValue(Index, SelectValue);
```
This breaks for repeated predecessors

tsymalla commented on llvm/lib/Passes/PassRegistry.def:
```diff
@@ -552,6 +552,7 @@ FUNCTION_PASS_WITH_PARAMS(
     "no-forward-switch-cond;forward-switch-cond;no-switch-range-to-icmp;"
     "switch-range-to-icmp;no-switch-to-lookup;switch-to-lookup;no-keep-loops;"
     "keep-loops;no-hoist-common-insts;hoist-common-insts;no-sink-common-insts;"
+    "switch-to-select"
```
Missing `;` at the end?

tsymalla commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6166,6 +6166,279 @@ static bool trySwitchToSelect(SwitchInst *SI, IRBuilder<> &Builder,
+    bool IsDefault = false) {
+  BasicBlock *Pred = SI->getParent();
+  int NumOfInsts = 0;
+  // Check if there is only one instruction per block, excepet from default
```
Typo: excepet

tsymalla commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6166,6 +6166,279 @@ static bool trySwitchToSelect(SwitchInst *SI, IRBuilder<> &Builder,
+          if (Instruction *I = dyn_cast<Instruction>(User))
+            if (I->getParent() == CaseDest)
+              continue;
+          if (PHINode *Phi = dyn_cast<PHINode>(User))
```
This could have some parentheses `{` and `}`, and you can pull the nested if into the top if via `;`

tsymalla commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6166,6 +6166,279 @@ static bool trySwitchToSelect(SwitchInst *SI, IRBuilder<> &Builder,
+    Res.push_back(std::make_pair(&PHI, Val));
+  }
+
+  return Res.size() > 0;
```
`!Res.empty()`

tsymalla commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6166,6 +6166,279 @@ static bool trySwitchToSelect(SwitchInst *SI, IRBuilder<> &Builder,
+  SwitchCaseResultVectorTy2 UniqueResults;
+  PHINodeToCaseEntryValueMapTy Map;
+
+  if (!initializeuniqueCasesWithoutConstants(
```
```suggestion
  if (!initializeUniqueCasesWithoutConstants(
```

tsymalla commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6166,6 +6166,279 @@ static bool trySwitchToSelect(SwitchInst *SI, IRBuilder<> &Builder,
+        Value *value = ValuePair.second;
+        if (constantInt == CaseVal)
+          if (!isa<Constant>(value))
+            IsSafeToRemove = false;
```
Shouldn't the loop break then?

Acim-Maravic committed:
[NFC][SimplifyCFG] Convert switch to cmp/select sequence

Added tests that will show changes

Acim-Maravic committed:
[SimplifyCFG] Convert switch to cmp/select sequence

tsymalla commented on llvm/lib/Transforms/Scalar/SimplifyCFGPass.cpp:
```diff
@@ -77,6 +77,9 @@ static cl::opt<bool> UserSinkCommonInsts(
     "sink-common-insts", cl::Hidden, cl::init(false),
     cl::desc("Sink common instructions (default = false)"));
 
+static cl::opt<bool> UserSwitchToSelect(
```
```suggestion
static cl::opt<bool> UserConvertSwitchToSelect(
```

tsymalla commented on llvm/lib/Transforms/Scalar/SimplifyCFGPass.cpp:
```diff
@@ -77,6 +77,9 @@ static cl::opt<bool> UserSinkCommonInsts(
     cl::desc("Sink common instructions (default = false)"));
 
+static cl::opt<bool> UserSwitchToSelect(
+    "switch-to-select", cl::Hidden, cl::init(false),
```
```suggestion
    "convert-switch-to-select", cl::Hidden, cl::init(false),
```

arsenm commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6197,6 +6197,268 @@ static bool trySwitchToSelect(SwitchInst *SI, IRBuilder<> &Builder,
+using SwitchCaseResultsTy2 = SmallVector<std::pair<PHINode *, Value *>, 4>;
+
+using PHINodeToCaseEntryValueMapTy =
+    std::map<PHINode *, SmallVector<std::pair<ConstantInt *, Value *>, 4>>;
```
Avoid using std::map

arsenm commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6197,6 +6197,268 @@ static bool trySwitchToSelect(SwitchInst *SI, IRBuilder<> &Builder,
+// The first field contains the phi node that generates a result of the switch
+// and the second field contains the value generated for a certain case in the
+// switch for that PHI.
+using SwitchCaseResultsTy2 = SmallVector<std::pair<PHINode *, Value *>, 4>;
```
Something better than a "2" suffix on the name

tsymalla commented on llvm/lib/Transforms/Utils/SimplifyCFG.cpp:
```diff
@@ -6197,6 +6197,268 @@ static bool trySwitchToSelect(SwitchInst *SI, IRBuilder<> &Builder,
+}
+
+static bool getCaseResultsWithoutConstants(
+    SwitchInst *SI, BasicBlock *CaseDest, BasicBlock **CommonDest,
```
```suggestion
    SwitchInst *SI, BasicBlock *CaseDest, BasicBlock *&CommonDest,
```

nikic reviewed (commented):
The PR description says *what* this does without explaining *why*. This does not look look like a generally (or even commonly) profitable transform to me, so more context on what you're trying to achieve here would be helpful.

tsymalla commented:
> The PR description says _what_ this does without explaining _why_. This does not look look like a generally (or even commonly) profitable transform to me, so more context on what you're trying to achieve here would be helpful.

Agree, I think the idea is to get rid of cases where we create lots of conditional code from switches instead of v_cndmasks. Will talk to you offline.

nikic commented:
@tsymalla If this is targeted at GPUs, maybe the SpeculativeExecution pass would be a better fit? That's the one that does the kind of aggressive speculation that's only really profitable for GPU targets. Currently it only handles branches, but it could probably be generalized to handle switches as well.
